File: desktop/frontend/src/lib/fsSegments.ts
// Mirror of internal/proto/fs_segments.go.
//
// FS payloads are segmented rather than a single JSON document so file
// bytes can ride raw inside an AEAD envelope instead of being base64'd
// into JSON — encoding/json already base64s []byte, so nesting an
// envelope in a JSON string would encode twice for 1.78x expansion.
//
//   payload := segment_count(1B) || segment*
//   segment := length(4B BE) || bytes
//
// Segment 0 is always plaintext JSON with the fields the relay routes
// on. Segment count doubles as the key-state signal: one segment means
// the peer is keyless, more means sealed.

const MAX_SEGMENT = 16 * 1024 * 1024;

export function encodeSegments(segments: Uint8Array[]): Uint8Array {
  if (segments.length === 0 || segments.length > 255) {
    throw new Error(`fs segments: bad segment count ${segments.length}`);
  }
  let total = 1;
  for (const s of segments) {
    if (s.length > MAX_SEGMENT) throw new Error("fs segments: segment too large");
    total += 4 + s.length;
  }
  const out = new Uint8Array(total);
  const view = new DataView(out.buffer);
  out[0] = segments.length;
  let offset = 1;
  for (const s of segments) {
    view.setUint32(offset, s.length, false);
    offset += 4;
    out.set(s, offset);
    offset += s.length;
  }
  return out;
}

/** decodeSegments returns null rather than throwing on malformed input:
 *  every caller sits on a frame-handling path where the right response to
 *  a bad payload is to drop it. Declared lengths must sum to exactly the
 *  remaining bytes, so truncation and padding both fail. */
export function decodeSegments(payload: Uint8Array): Uint8Array[] | null {
  if (payload.length < 1) return null;
  const count = payload[0];
  if (count === 0) return null;
  const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  const segments: Uint8Array[] = [];
  let offset = 1;
  for (let i = 0; i < count; i++) {
    if (offset + 4 > payload.length) return null;
    const n = view.getUint32(offset, false);
    if (n > MAX_SEGMENT) return null;
    offset += 4;
    if (offset + n > payload.length) return null;
    segments.push(payload.slice(offset, offset + n));
    offset += n;
  }
  if (offset !== payload.length) return null;
  return segments;
}

File: internal/proto/fs_segments.go
package proto

import (
	"encoding/binary"
	"errors"
	"fmt"
)

// FS frame payloads are segmented rather than a single JSON document so
// file bytes can travel raw inside an AEAD envelope instead of being
// base64'd into JSON. Segment 0 is always plaintext JSON carrying the
// fields the relay routes on; later segments are sealed envelopes.
//
//	payload := segment_count(1B) || segment*
//	segment := length(4B BE) || bytes
//
// The other three sealed sites (META / SessionInfo / CommandEvent)
// base64 an envelope into a JSON string field. That is right for their
// scale — tens of bytes of metadata — but would base64 file contents a
// second time on top of the encoding encoding/json already applies to
// []byte, for 1.78x expansion. Segments keep it at ~1.0x.
var (
	ErrFSSegmentsMalformed = errors.New("proto: malformed FS segments")
	ErrFSSegmentTooLarge   = errors.New("proto: FS segment exceeds max payload")
)

// EncodeSegments frames segments into a single payload. At least one
// segment is required; segment 0 is by convention the plaintext head.
func EncodeSegments(segments [][]byte) ([]byte, error) {
	if len(segments) == 0 {
		return nil, fmt.Errorf("%w: need at least one segment", ErrFSSegmentsMalformed)
	}
	if len(segments) > 255 {
		return nil, fmt.Errorf("%w: %d segments", ErrFSSegmentsMalformed, len(segments))
	}
	total := 1
	for _, s := range segments {
		if len(s) > maxPayload {
			return nil, ErrFSSegmentTooLarge
		}
		total += 4 + len(s)
	}
	out := make([]byte, 0, total)
	out = append(out, byte(len(segments)))
	for _, s := range segments {
		var lenBuf [4]byte
		binary.BigEndian.PutUint32(lenBuf[:], uint32(len(s)))
		out = append(out, lenBuf[:]...)
		out = append(out, s...)
	}
	return out, nil
}

// DecodeSegments is the inverse of EncodeSegments. It rejects declared
// lengths that do not sum to exactly the remaining bytes, so a truncated
// or padded payload fails loudly rather than yielding partial segments.
func DecodeSegments(payload []byte) ([][]byte, error) {
	if len(payload) < 1 {
		return nil, fmt.Errorf("%w: empty payload", ErrFSSegmentsMalformed)
	}
	count := int(payload[0])
	if count == 0 {
		return nil, fmt.Errorf("%w: zero segments", ErrFSSegmentsMalformed)
	}
	rest := payload[1:]
	segments := make([][]byte, 0, count)
	for i := 0; i < count; i++ {
		if len(rest) < 4 {
			return nil, fmt.Errorf("%w: truncated length prefix at segment %d", ErrFSSegmentsMalformed, i)
		}
		n := int(binary.BigEndian.Uint32(rest[:4]))
		if n > maxPayload {
			return nil, ErrFSSegmentTooLarge
		}
		rest = rest[4:]
		if len(rest) < n {
			return nil, fmt.Errorf("%w: truncated body at segment %d", ErrFSSegmentsMalformed, i)
		}
		seg := make([]byte, n)
		copy(seg, rest[:n])
		segments = append(segments, seg)
		rest = rest[n:]
	}
	if len(rest) != 0 {
		return nil, fmt.Errorf("%w: %d trailing bytes", ErrFSSegmentsMalformed, len(rest))
	}
	return segments, nil
}
